flash/data/batch.py

```diff
@@ -0,0 +1,186 @@
+        if self.assert_contains_tensor:
+            if not _contains_any_tensor(sample):
+                raise MisconfigurationException(
+                    "When ``per_sample_to_tensor_transform`` is overriden, "
+                    "``DataPipeline`` expects the outputs to be ``tensors``"
+                )
+        sample = self.per_sample_post_tensor_transform(sample)
+        return sample
+
+    def __str__(self) -> str:
+        repr_str = f'{self.__class__.__name__}:'
+        repr_str += f'\n\t\t(per_sample_pre_tensor_transform): {repr(self.per_sample_pre_tensor_transform)}'
+        repr_str += f'\n\t\t(per_sample_to_tensor_transform): {repr(self.per_sample_to_tensor_transform)}'
+        repr_str += f'\n\t\t(per_sample_post_tensor_transform): {repr(self.per_sample_post_tensor_transform)}'
+        repr_str += f'\n\t\t(assert_contains_tensor): {repr(self.assert_contains_tensor)}'
+        return repr_str
+
+
+class _PreProcessor(torch.nn.Module):
+    """
+        This class is used to encapsultate the following functions of a Preprocess Object:
+        Inside a worker:
+            per_sample_transform: Function to transform an individual sample
+                Inside a worker, it is actually make of 3 functions:
+                    * per_sample_pre_tensor_transform
+                    * per_sample_to_tensor_transform
+                    * per_sample_post_tensor_transform
+            collate: Function to merge sample into a batch
+            per_batch_transform: Function to transform an individual batch
+                * per_batch_transform
+
+        Inside main process:
+            per_sample_transform: Function to transform an individual sample
+                * per_sample_transform_on_device
+            collate: Function to merge sample into a batch
+            per_batch_transform: Function to transform an individual batch
+                * per_batch_transform_on_device
+    """
+
+    def __init__(
+        self,
+        collate_fn: Callable,
+        per_sample_transform: Union[Callable, _Sequential],
+        per_batch_transform: Callable,
+        stage: Optional[RunningStage] = None,
+        apply_per_sample_transform: bool = True,
+    ):
+        super().__init__()
+        self.collate_fn = convert_to_modules(collate_fn)
+        self.per_sample_transform = convert_to_modules(per_sample_transform)
+        self.per_batch_transform = convert_to_modules(per_batch_transform)
+        self.apply_per_sample_transform = apply_per_sample_transform
+        self.stage = stage
+
+    def forward(self, samples: Sequence[Any]):
+        if self.apply_per_sample_transform:
+            samples = [self.per_sample_transform(sample) for sample in samples]
+            samples = type(samples)(samples)
+            samples = self.collate_fn(samples)
+        samples = self.per_batch_transform(samples)
+        return samples
+
+    def __str__(self) -> str:
+        repr_str = '_PreProcessor:'
```

Comment:
```suggestion
        # todo: define repr function which would take object and string attributes to be shown
        repr_str = '_PreProcessor:'
```
this is more later work, just adding note